Add simplification

cheginit opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

cheginit commented:
@urschrei, please confirm that you will be a maintainer for `simplification` on `conda-forge`. (Addresses [simplification/issues#44](https://github.com/urschrei/simplification/issues/44))

cheginit commented:
@conda-forge/help-python-c I'd appreciate it it you could help with debugging the issue with the build?

urschrei commented:
I don't think your recipe will be able to build in its current form, as the simplification source you pull in from PyPI doesn't include `librdp`. If you have a look [at how the simplification CI build works](https://github.com/urschrei/simplification/blob/master/.github/workflows/wheels.yml#L72-L102), the correct platform library is downloaded from [github](https://github.com/urschrei/rdp/releases/tag/v0.13.0) and unpacked into the source at wheel build time.

cheginit commented:
@urschrei I see. I think in that case, we should have a build script that takes care of that. Do you have something for that part? I pinged the `conda-forge` helper, hopefully they can help with this too.

cheginit commented:
@urschrei, there's also [this](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=948690&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=884) error in the logs. Can you please check if that's related to what you're saying, or it's something else?

urschrei commented:
> @urschrei, there's also [this](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=948690&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=884) error in the logs. Can you please check if that's related to what you're saying, or it's something else?

That's the same issue: `header.h` is pulled in from the rdp github release source I lined above. I don't have a build script per se as I use Github Actions; the link I posted above has the Bash invocation I use once the action has downloaded and un-tar'd the source.

cheginit commented:
@urschrei, alright, I figured it out! I added two build scripts for handling compilation on Unix and Windows. I also fixed the deps for the build and host blocks. The only thing is that `rdb` version needs to be set manually in the `meta.yml` file, since using `curl` to get the latest version was prone to failing for some reason.

@conda-forge/help-python-c Only the Windows build fails. It appears that building the package finishes successfully, but packaging it fails due to not finding `rdp.dll`. This is the error:

```console
['  ERROR (simplification,Lib/site-packages/simplification/cutil.cp310-win_amd64.pyd): $RPATH/rdp.dll not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?']
```

I even tried adding the DLL to the DSO whitelist, but it didn't seem to help. Can please help with debugging this?

cheginit commented:
@conda-forge/help-python-c I figured out a solution and now the package is ready for review, thanks!

xhochy commented on recipes/simplification/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+    - python
+    - setuptools >=45
+    - setuptools-scm >=6.2
+    - numpy >=1.11.0
```
Please keep this unpinned so the global pin is taken
```suggestion
    - numpy
```

xhochy commented on recipes/simplification/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+  build:
+    - python  # [build_platform != target_platform]
+    - cross-python_{{ target_platform }}  # [build_platform != target_platform]
+    - setuptools >=45
+    - setuptools-scm >=6.2
+    - numpy >=1.11.0
+    - cython >=0.29.0
```
Thse should only be in host
```suggestion
```

xhochy commented on recipes/simplification/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+  script_env:
+    - RDPTAG={{ rdp_version }}
+  missing_dso_whitelist:  # [win]
+    - "*/rdp.dll"  # [win]
```
Where is this coming from?

cheginit replied:
The windows build, for some reason, during packaging (not building) cannot find this file, so I added it to the whitelist.

xhochy commented on recipes/simplification/build.sh:
```diff
@@ -0,0 +1,19 @@
+	FILENAME="rdp-${TAG}-x86_64-apple-darwin.tar.gz"
+fi
+
+URL="https://github.com/urschrei/rdp/releases/download/${TAG}/${FILENAME}"
+curl -L $URL -o $FILENAME
+tar -xvf $FILENAME -C src/simplification
```
This should be built from source and be in a separate conda package.

cheginit replied:
`rdp` is a Rust library, can you please let me know how can we deal with this? @urschrei, who is the developer of both packages, builds `simplification` for pypi using a similar approach.

xhochy replied:
We have several Rust packages on conda-forge, e.g. see the just merged https://github.com/conda-forge/staged-recipes/pull/25590

cheginit replied:
Thanks for the link! I looked into it and realized there's a difference between `rdp` and the example Rust packages: Building `rdp` from source only generates one file (`librdp.dylib` for macOS, `librdp.so` for Linux, or `rdp.dll` for Windows) and stores it in the `$SRC_DIR/target/release/`. It doesn't produce any binaries.

Would it be possible, instead of creating a new `conda-forge` package, I modify the build script to generate the library file from source by downloading the source code, for example, from `https://github.com/urschrei/rdp/archive/refs/tags/v0.13.0.tar.gz`, build it with `CARGO_INCREMENTAL="0" cargo build --release`, and finally copy the generated platform specific library to `src/simplification`?

xhochy commented on conda-forge.yml:
```diff
@@ -3,3 +3,7 @@ azure:
   settings_win:
     pool:
       vmImage: windows-2022
+build_platform:
```
Don't touch this file, please revert the changes here.

cheginit replied:
Ok, I will. How should I add `osm-arm64` build?

cheginit commented:
@xhochy I'd appreciate it if you check the latest changes that I made to address the issues that you raised.

ForrestQuant commented on recipes/simplification/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+    - pip
+  run:
+    - python
+    - numpy >=1.11.0
```
This will come in via `run_exports` and should not be listed explicitly.
```suggestion
```

ForrestQuant commented on recipes/simplification/build.sh:
```diff
@@ -0,0 +1,56 @@
+    cp target/${TARGET}/release/librdp.dylib ../
+    ;;
+  osx-arm64)
+	xcodebuild -showsdks
+	SDKROOT=$(xcrun -sdk macosx14.0 --show-sdk-path)
```
This should be set via `c_stdlib_version` in `conda_build_config.yaml` (you can do this on a feedstock and rerendering)
```suggestion
	xcodebuild -showsdks
	SDKROOT=$(xcrun -sdk macosx14.0 --show-sdk-path)
```

cheginit replied:
I will.

ForrestQuant commented on recipes/simplification/build.sh:
```diff
@@ -0,0 +1,56 @@
+    ;;
+  osx-64)
+  	TARGET=x86_64-apple-darwin
+	export MACOSX_DEPLOYMENT_TARGET=10.9
```
See below
```suggestion
```

ForrestQuant commented on recipes/simplification/build.sh:
```diff
@@ -0,0 +1,56 @@
+	xcodebuild -showsdks
+	SDKROOT=$(xcrun -sdk macosx14.0 --show-sdk-path)
+  	TARGET=aarch64-apple-darwin
+	export MACOSX_DEPLOYMENT_TARGET=11.0
```
See above
```suggestion
```

xhochy commented:
Used the wrong account to comment, sorry. Review still applies ;)

cheginit commented:
@xhochy Awesome, thanks! I committed your changes and all builds finished successfully.
